common/models/model-definition.js

```diff
@@ -75,10 +75,19 @@ function ready(ModelDefinition) {
     assert(relation.embed && relation.embed.as, 'embed requires "as"');
     switch (relation.embed.as) {
       case 'object':
-        assert(relation.embed.key, 'embed as object requires "key"');
+        assert(relation.embed.key || relation.embed.keySetter,
+          'embed as object requires "key" or "keySetter"');
         result = {};
         relatedData.forEach(function(related) {
-          var key = related[relation.embed.key];
+          var Definition = app.models[relation.model];
+          var key;
+          if (relation.embed.key) {
+            key = related[relation.embed.key];
+          }
+          var keySetter = relation.embed.keySetter;
```

Comment:
I think this should be `keyGetter`, because we want to build a key from the data stored in our domain entity, shouldn't it?

The idea in my [comment](https://github.com/strongloop/loopback-workspace/pull/273#discussion_r62665458) was to allow two functions: `keyGetter` to get the key from domain data (cache entry), `keySetter` to set the key in the domain data (cache entry).

Now if we can implement this in such way that only a single function is needed, then that's great! Just use a different name for the single function, I think the original `keyFn` is good enough.

However, I am little bit concerned that the approach of a single method may not be future proof - what if the cached data is very different from the JSON data and the algorithms for get and set are thus different too?

@0candy thoughts?